streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -108,66 +94,58 @@ public Admin adminClient() {
         return adminClient;
     }
 
-    void createTasks(final Collection<TopicPartition> assignment) {
-        if (consumer == null) {
-            throw new IllegalStateException(logPrefix + "consumer has not been initialized while adding stream tasks. This should not happen.");
-        }
-
-        if (!assignment.isEmpty() && !assignedActiveTasks.isEmpty()) {
-            resumeSuspended(assignment);
-        }
-        if (!addedActiveTasks.isEmpty()) {
-            addNewActiveTasks(addedActiveTasks);
-        }
-        if (!addedStandbyTasks.isEmpty()) {
-            addNewStandbyTasks(addedStandbyTasks);
-        }
-
-        // need to clear restore consumer if it was reading standbys but we have active tasks that may need restoring
-        if (!addedActiveTasks.isEmpty() && restoreConsumerAssignedStandbys) {
-            restoreConsumer.unsubscribe();
-            restoreConsumerAssignedStandbys = false;
-        }
-
-        // Pause all the new partitions until the underlying state store is ready for all the active tasks.
-        log.debug("Pausing all active task partitions until the underlying state stores are ready");
-        pausePartitions();
-    }
-
-    private void resumeSuspended(final Collection<TopicPartition> assignment) {
-        final Set<TaskId> suspendedTasks = partitionsToTaskSet(assignment);
-        suspendedTasks.removeAll(addedActiveTasks.keySet());
-
-        log.debug("Suspended tasks to be resumed: {}", suspendedTasks);
-
-        for (final TaskId taskId : suspendedTasks) {
-            final Set<TopicPartition> partitions = assignedActiveTasks.get(taskId);
-            try {
-                if (!active.maybeResumeSuspendedTask(taskId, partitions)) {
-                    // recreate if resuming the suspended task failed because the associated partitions changed
-                    addedActiveTasks.put(taskId, partitions);
+    public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
```

Comment:
This is a meta comment about where we should handle task ownership changes as source of truth:

1) listener#onPartitionsAssigned is triggered at the end of every rebalance where only new-delta is passed in; also assignor#onAssignment is triggered at the end of every rebalance where the complete list of active/standby are available.

2) listener#onPartitionsRevoked is not necessarily triggered on each rebalance; HOWEVER it may also be triggered outside of a rebalance, when we are re-subscribing / closing and some owned partitions are no longer subscribed. For streams this should not happen though since we never re-subscribe; on upon closing we should have closed the task manager and hence closed all tasks already before closing the consumer so when onRevoked is triggered there should be no tasks left already.

3) listener#onPartitionsLost is not triggered on each rebalance; HOWEVER it may be triggered whenever the consumer discovered a re-join group needed indicating that all tasks are lost. In Streams it can happen, and although it may also be notified from other call path like stream task committing getting an error, it is not guaranteed, i.e. it may not be notified elsewhere.

So there are two cases that Streams should consider as the source-of-truth task ownership changes: 1)  assignor#onAssignment, and 3) listener#onPartitionsLost.

And in 2) listener#onPartitionsRevoked there should be no tasks ownership change but ONLY cause active tasks to be suspended --- and also we should consider removing the suspension / resumption in the future with KIP-429.

Unfortunately it means that we need to rely on two classes, both the assignor and the listener to guide the task manager on ownership and task-state changes:

a) from assignor#onAssignment, we can create tasks, close tasks (for both active and standby).
b) from listener#onPartitionsRevoked, we only suspend active tasks.
c) from listener#onPartitionsLost, we close-dirty active tasks and leave standby tasks unchanged (? open for debate).
d) we do not need to do anything with listener#onPartitionsAssigned.

From the current PR it seems you've already got it all, except in c) we are closing all active and standby as dirty. My rationale is that after re-join we are likely to get back those standby tasks, and even if we do not we will be noticed on the next rebalance and close standby tasks anyways, so we only keep them for a bit longer (and semantically it is okay to temporarily keep more standby tasks at a given time than configured number). WDYT?

Also cc @ableegoldman to double check and the above analysis related to KIP-429.

Reply:
Thanks @guozhangwang , that sounds like exactly the thoughts that were in my own head :) .

Regarding (c), it sounds like a good argument. I can make that change.